File: shared/essentials/web.py
"""Essentials-book web API helpers built on shared web primitives."""

from __future__ import annotations

from typing import Optional, Union

from fastapi import Body, FastAPI, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from shared.utils import (
    normalize_response_text,
    parse_structured_json_response,
)
from shared.llm_models import get_identifier_mappings
from shared.essentials.utils import selected_model_context


class SharedQueryRequest(BaseModel):
    topic: str = Field(..., description="Prompt or question")
    provider: Optional[Union[str, int]] = None
    model_identifier: Optional[Union[str, int]] = None
    modelIdentifier: Optional[Union[str, int]] = None
    template: str = "{topic}"
    max_tokens: int = 1000
    temperature: float = 0.7
    session_id: Optional[str] = "default"
    sessionId: Optional[str] = None


class SharedQueryAllRequest(BaseModel):
    topic: str
    template: str = "{topic}"
    max_tokens: int = 1000
    temperature: float = 0.7
    session_id: Optional[str] = "default"
    sessionId: Optional[str] = None


class SharedResetMemoryRequest(BaseModel):
    provider: Optional[Union[str, int]] = None
    session_id: Optional[str] = None
    sessionId: Optional[str] = None


from shared.web import (
    available_model_identifiers,
    build_manager,
    resolve_session_id,
    model_payload,
    model_payloads,
    model_identifiers_for_provider,
    normalize_model_identifier_input,
    normalize_provider_input,
    provider_payload,
    provider_selection_map,
    result_is_success,
    run_manager_in_thread,
    run_uvicorn_app,
    stream_text_sse,
    supports_coagent,
    supports_memory,
    supports_memory_retrieval,
    supports_session_memory,
    to_sse_line,
)


def _parse_structured_raw_response(raw_response):
    if raw_response is None:
        return None
    try:
        parsed = parse_structured_json_response(raw_response)
        return parsed if isinstance(parsed, dict) else None
    except Exception:
        return None


def _extract_answer_text(response_obj, raw_response):
    if isinstance(response_obj, dict):
        answer = response_obj.get("answer") or response_obj.get("distilled") or response_obj.get("summary")
        if isinstance(answer, str) and answer.strip():
            return answer
    return normalize_response_text(raw_response)


def recover_structured_parse_error(result: dict) -> dict:
    if not isinstance(result, dict):
        return result

    raw_response = result.get("raw_response")
    parsed_response = _parse_structured_raw_response(raw_response)

    if result.get("success"):
        if parsed_response:
            response = result.get("response")
            metadata_notes = None
            if isinstance(response, dict):
                metadata_notes = (response.get("metadata") or {}).get("notes")
            if isinstance(metadata_notes, str) and metadata_notes.startswith("Failed to parse structured JSON response"):
                return {**result, "response": parsed_response, "raw_answer": _extract_answer_text(parsed_response, raw_response)}
        return result

    error_message = str(result.get("error") or "")
    if "Failed to parse structured JSON response" not in error_message or not raw_response:
        return result

    recovered_response = parsed_response or {
        "answer": normalize_response_text(raw_response),
        "distilled": normalize_response_text(raw_response),
        "metadata": {"confidence": "low", "notes": error_message},
    }

    return {**result, "success": True, "error": None, "response": recovered_response, "raw_answer": _extract_answer_text(recovered_response, raw_response)}



def result_value(result: dict, *keys, default=None):
    for key in keys:
        if isinstance(result, dict) and key in result and result.get(key) is not None:
            return result.get(key)
    return default


def is_corrupt_history_error(exc: Exception) -> bool:
    message = str(exc)
    return any(signal in message for signal in ("string indices must be integers", "Expecting value", "JSON"))


def ask_question_with_recovery(manager, args: dict, session_id: Optional[str]):
    try:
        return manager.ask_question(**args)
    except Exception as exc:
        if not supports_session_memory(manager) or not is_corrupt_history_error(exc):
            raise
        provider = args.get("provider")
        try:
            manager.reset_memory(provider, session_id)
        except Exception:
            raise exc
        return manager.ask_question(**args)


def build_query_args(manager, request: SharedQueryRequest) -> tuple[dict, str, Optional[str]]:
    effective_session_id = resolve_session_id(request.sessionId, request.session_id)
    requested_model_identifier = request.modelIdentifier if request.modelIdentifier is not None else request.model_identifier
    model_identifier = normalize_model_identifier_input(manager, requested_model_identifier)
    provider_input = request.provider
    if model_identifier is None:
        model_identifier = normalize_model_identifier_input(manager, provider_input)
    if model_identifier:
        provider = get_identifier_mappings()[model_identifier].provider
    else:
        provider = normalize_provider_input(manager, provider_input)
    args = {
        "topic": request.topic,
        "provider": provider,
        "template": request.template,
        "max_tokens": request.max_tokens,
        "temperature": request.temperature,
    }
    if supports_session_memory(manager):
        args["session_id"] = effective_session_id
    return args, effective_session_id, model_identifier


async def execute_manager_query(manager, args: dict, session_id: Optional[str], model_identifier: Optional[str] = None) -> dict:
    def run_query():
        with selected_model_context(model_identifier):
            return ask_question_with_recovery(manager, args, session_id)

    return recover_structured_parse_error(await run_manager_in_thread(run_query))


def query_error_message(result: dict) -> str:
    if not isinstance(result, dict):
        return "Manager returned a non-dict response"
    return result.get("error", "Query failed")


def serialize_query_response(manager, request: SharedQueryRequest, result: dict, session_id: str, model_identifier: Optional[str] = None) -> dict:
    raw_response = result.get("response")
    content = raw_response if isinstance(raw_response, (dict, list)) else normalize_response_text(raw_response)
    return {
        "success": True,
        "framework": manager.framework,
        "topic": request.topic,
        "selected_provider": request.provider or model_identifier,
        "provider": result_value(result, "provider", default="unknown"),
        "model": result_value(result, "model", default=""),
        "model_identifier": model_identifier,
        "response": content,
        "parameters": {
            "temperature": result_value(result, "temperature"),
            "max_tokens": result_value(result, "max_tokens", "maxTokens"),
            "template": request.template,
        },
        "prompt": result_value(result, "prompt", default=request.template.format(topic=request.topic)),
        "session_id": result_value(result, "session_id", "sessionId", default=session_id),
    }


def serialize_done_event(result: dict, raw_response, session_id: str) -> dict:
    return {
        "type": "done",
        "provider": result_value(result, "provider"),
        "model": result_value(result, "model"),
        "model_identifier": result_value(result, "model_identifier"),
        "response": raw_response if isinstance(raw_response, dict) else None,
        "token_usage": result_value(result, "token_usage", "tokenUsage"),
        "session_id": result_value(result, "session_id", "sessionId", default=session_id),
    }


def create_web_api(manager_class):
    app = FastAPI(title="LLM Service API", version="1.0.0", description="Universal API for LLM framework testing")
    app.add_middleware(CORSMiddleware, allow_origins=["*"], allow_credentials=True, allow_methods=["*"], allow_headers=["*"])
    manager = build_manager(manager_class)

    @app.get('/')
    async def get_status():
        available = manager.get_available_providers()
        return {
            "framework": manager.framework,
            "available_providers": available,
            "available_provider_details": [provider_payload(provider, manager) for provider in available],
            "available_model_details": model_payloads(manager),
            "total_available": len(available),
            "total_models_available": len(model_payloads(manager)),
            "initialization_status": manager.initialization_messages,
            "status": "healthy" if available else "no_providers",
        }

    @app.get('/providers')
    async def get_providers():
        providers = manager.get_available_providers()
        models = model_payloads(manager)
        active = models[0]["name"] if models else None
        return {
            "framework": manager.framework,
            "providers": models,
            "provider_groups": [provider_payload(provider, manager) for provider in providers],
            "available_providers": providers,
            "models": models,
            "count": len(models),
            "provider_count": len(providers),
            "model_count": len(models),
            "active_provider": active,
        }

    @app.get('/capabilities')
    async def get_capabilities():
        return {"framework": manager.framework, "streaming": True, "memory": supports_memory(manager), "memory_retrieval": supports_memory_retrieval(manager), "coagent": supports_coagent(manager)}

    @app.post('/query')
    async def query_single(request: SharedQueryRequest):
        try:
            args, effective_session_id, model_identifier = build_query_args(manager, request)
            result = await execute_manager_query(manager, args, effective_session_id, model_identifier)
            if isinstance(result, dict) and model_identifier:
                result["model_identifier"] = model_identifier
            if not isinstance(result, dict):
                raise HTTPException(status_code=500, detail=query_error_message(result))
            if not result_is_success(result):
                raise HTTPException(status_code=400, detail=query_error_message(result))
            return serialize_query_response(manager, request, result, effective_session_id, model_identifier)
        except HTTPException:
            raise
        except Exception as exc:
            raise HTTPException(status_code=500, detail=str(exc)) from exc

    @app.post('/query-stream')
    async def query_stream(request: SharedQueryRequest):
        async def stream_events():
            try:
                args, effective_session_id, model_identifier = build_query_args(manager, request)
                result = await execute_manager_query(manager, args, effective_session_id, model_identifier)
                if isinstance(result, dict) and model_identifier:
                    result["model_identifier"] = model_identifier
                if not isinstance(result, dict) or not result_is_success(result):
                    yield to_sse_line({"type": "error", "error": query_error_message(result)})
                    return
                raw_response = result.get('response')
                async for event in stream_text_sse(normalize_response_text(raw_response), delay_seconds=0.03):
                    yield event
                yield to_sse_line(serialize_done_event(result, raw_response, effective_session_id))
            except Exception as exc:
                yield to_sse_line({"type": "error", "error": str(exc)})
        return StreamingResponse(stream_events(), media_type='text/event-stream')

    @app.get('/history')
    async def get_history(provider: str = 'openai', session_id: str = 'default'):
        if not supports_session_memory(manager):
            raise HTTPException(status_code=400, detail='Session memory not supported by this manager')
        return manager.get_history(provider, session_id)

    @app.post('/reset-memory')
    async def reset_memory(request: Optional[SharedResetMemoryRequest] = Body(None), provider: Optional[str] = None, session_id: Optional[str] = None):
        if not supports_session_memory(manager):
            raise HTTPException(status_code=400, detail='Session memory not supported by this manager')
        body_provider = request.provider if request else None
        body_session_id = (request.sessionId or request.session_id) if request else None
        effective_provider = body_provider if body_provider is not None else provider
        effective_session_id = body_session_id if body_session_id is not None else session_id
        return manager.reset_memory(effective_provider, effective_session_id)

    return app


def run_web_server(manager_class, host: str = '0.0.0.0', port: int = 8000):
    app = create_web_api(manager_class)
    try:
        framework_name = build_manager(manager_class).framework
    except Exception:
        framework_name = 'Unknown'
    run_uvicorn_app(app, framework_name, host=host, port=port)

File: shared/essentials/web.mjs
import {
  getAllProviders,
  getDefaultModelDetails,
  normalizeResponseText,
  parseStructuredJsonResponse,
  sortProvidersByDisplayOrder,
} from "../utils.mjs";
import { ALL_MODEL_IDENTIFIERS, getIdentifierMappings } from "../llm_models.mjs";
import { withSelectedModelIdentifier } from "./utils.mjs";
import {
  buildManager,
  captureConsoleOutputAsync,
  createExpressApp,
  resolveSessionId,
  resultIsSuccess,
  streamTextSse,
  supportsCoagent,
  supportsMemory,
  supportsMemoryRetrieval,
  supportsSessionMemory,
  toSseLine,
} from "../web.mjs";

export function parseStructuredRawResponse(rawResponse) {
  if (typeof rawResponse === "undefined" || rawResponse === null) return null;
  try {
    const parsed = parseStructuredJsonResponse(rawResponse);
    return parsed && typeof parsed === "object" && !Array.isArray(parsed) ? parsed : null;
  } catch {
    return null;
  }
}

export function extractAnswerText(responseObj, rawResponse) {
  if (responseObj && typeof responseObj === "object") {
    const answer = responseObj.answer || responseObj.distilled || responseObj.summary;
    if (typeof answer === "string" && answer.trim()) return answer;
  }
  return normalizeResponseText(rawResponse);
}

export function recoverStructuredParseError(result) {
  if (!result) return result;
  const rawResponse = result.rawResponse;
  const parsedResponse = parseStructuredRawResponse(rawResponse);
  if (result.success) {
    if (parsedResponse && result.response && typeof result.response === "object") {
      const metadataNotes = result.response?.metadata?.notes;
      if (typeof metadataNotes === "string" && metadataNotes.startsWith("Failed to parse structured JSON response")) {
        return { ...result, response: parsedResponse, rawAnswer: extractAnswerText(parsedResponse, rawResponse) };
      }
    }
    return result;
  }
  const errorMessage = String(result.error || "");
  if (!errorMessage.includes("Failed to parse structured JSON response") || !rawResponse) return result;
  const normalized = normalizeResponseText(rawResponse);
  const recoveredResponse = parsedResponse || { answer: normalized, distilled: normalized, metadata: { confidence: "low", notes: errorMessage } };
  return { ...result, success: true, error: null, response: recoveredResponse, rawAnswer: extractAnswerText(recoveredResponse, rawResponse) };
}

export function providerSelectionMap(manager) {
  const sortedProviders = sortProvidersByDisplayOrder(manager.getAvailableProviders());
  return Object.fromEntries(sortedProviders.map((provider, index) => [String(index + 1), provider]));
}

export function availableModelIdentifiers(manager) {
  const mappings = getIdentifierMappings();
  const availableProviders = new Set(manager.getAvailableProviders());
  return ALL_MODEL_IDENTIFIERS.filter((identifier) => mappings[identifier] && availableProviders.has(mappings[identifier].provider));
}

export function modelPayload(modelIdentifier, manager, idx = null) {
  const config = getIdentifierMappings()[modelIdentifier];
  let resolvedIdx = idx;
  if (resolvedIdx === null || typeof resolvedIdx === "undefined") {
    const availableIndex = availableModelIdentifiers(manager).indexOf(modelIdentifier);
    resolvedIdx = availableIndex >= 0 ? availableIndex + 1 : ALL_MODEL_IDENTIFIERS.indexOf(modelIdentifier) + 1;
  }
  const canonicalName = `${config.provider}:${config.model}`;
  return {
    id: String(resolvedIdx),
    name: canonicalName,
    display_name: `${config.provider.charAt(0).toUpperCase()}${config.provider.slice(1)} (${modelIdentifier})`,
    provider: config.provider,
    default_model: config.model,
    model: config.model,
    model_identifier: modelIdentifier,
    default_model_identifier: modelIdentifier,
    model_tier: config.tier,
    default_model_tier: config.tier,
    strengths: [...(config.strengths || [])],
    status: manager.initializationMessages[config.provider] || "Unknown",
    framework: manager.framework || "unknown",
  };
}

export function modelPayloads(manager) {
  return availableModelIdentifiers(manager).map((identifier, index) => modelPayload(identifier, manager, index + 1));
}

export function modelIdentifiersForProvider(provider) {
  const mappings = getIdentifierMappings();
  return ALL_MODEL_IDENTIFIERS.filter((identifier) => mappings[identifier] && mappings[identifier].provider === provider);
}

export function normalizeProviderInput(manager, provider) {
  if (provider === null || typeof provider === "undefined") return null;
  const providerMap = providerSelectionMap(manager);
  const available = new Set(manager.getAvailableProviders().map((p) => String(p).toLowerCase()));
  const configured = new Set(getAllProviders().map((p) => String(p).toLowerCase()));
  if (typeof provider === "number") return providerMap[String(provider)] ?? null;
  const candidate = String(provider).trim();
  if (!candidate) return null;
  if (candidate in providerMap) return providerMap[candidate];
  const lowered = candidate.toLowerCase();
  if (available.has(lowered) || configured.has(lowered)) return lowered;
  return candidate;
}

export function normalizeModelIdentifierInput(manager, modelIdentifier) {
  if (modelIdentifier === null || typeof modelIdentifier === "undefined") return null;
  const mappings = getIdentifierMappings();
  const payloads = modelPayloads(manager);
  const modelMap = Object.fromEntries(payloads.map((payload) => [payload.id, payload.model_identifier]));
  const canonicalMap = Object.fromEntries(payloads.map((payload) => [payload.name.toLowerCase(), payload.model_identifier]));
  if (typeof modelIdentifier === "number") return modelMap[String(modelIdentifier)] ?? null;
  const candidate = String(modelIdentifier).trim();
  if (!candidate) return null;
  if (candidate in modelMap) return modelMap[candidate];
  if (candidate in mappings) return candidate;
  const lowered = candidate.toLowerCase();
  if (lowered in canonicalMap) return canonicalMap[lowered];
  return Object.keys(mappings).find((identifier) => {
    const config = mappings[identifier];
    return identifier.toLowerCase() === lowered || config.model.toLowerCase() === lowered;
  }) || null;
}

export function buildProviderPayload(provider, manager) {
  const details = getDefaultModelDetails(provider);
  const modelIdentifiers = modelIdentifiersForProvider(provider);
  return {
    name: provider,
    display_name: details.displayName,
    provider: details.canonicalProvider,
    default_model: details.defaultModel,
    default_model_identifier: details.defaultModelIdentifier,
    default_model_tier: details.defaultModelTier,
    models: modelIdentifiers.map((identifier) => modelPayload(identifier, manager)),
    model_identifiers: modelIdentifiers,
    status: manager.initializationMessages[provider] || "Unknown",
  };
}

export async function askQuestionWithSession(manager, topic, provider, template, maxTokens, temperature, sessionId) {
  if (supportsSessionMemory(manager)) {
    return manager.askQuestion(topic, provider, template, maxTokens, temperature, sessionId);
  }
  return manager.askQuestion(topic, provider, template, maxTokens, temperature);
}

export function buildQueryPayload(manager, body = {}) {
  const {
    topic,
    provider = null,
    model_identifier = null,
    modelIdentifier = null,
    template = "{topic}",
    max_tokens = 1000,
    temperature = 0.7,
    session_id = "default",
    sessionId = null,
  } = body;
  const effectiveSessionId = resolveSessionId(sessionId, session_id);
  let selectedModelIdentifier = normalizeModelIdentifierInput(manager, modelIdentifier ?? model_identifier);
  if (!selectedModelIdentifier) selectedModelIdentifier = normalizeModelIdentifierInput(manager, provider);
  const resolvedProvider = selectedModelIdentifier
    ? getIdentifierMappings()[selectedModelIdentifier].provider
    : normalizeProviderInput(manager, provider);
  return {
    topic,
    provider: resolvedProvider,
    modelIdentifier: selectedModelIdentifier,
    template,
    maxTokens: max_tokens,
    temperature,
    sessionId: effectiveSessionId,
  };
}

export async function executeManagerQuery(manager, queryPayload) {
  const { topic, provider, template, maxTokens, temperature, sessionId, modelIdentifier } = queryPayload;
  return captureConsoleOutputAsync(async () => recoverStructuredParseError(
    await withSelectedModelIdentifier(modelIdentifier, async () => askQuestionWithSession(manager, topic, provider, template, maxTokens, temperature, sessionId)),
  ));
}

export function queryErrorPayload(result, logs = "") {
  return {
    error: result?.error || "Query failed",
    provider: result?.provider,
    debug: logs || null,
  };
}

export function serializeQueryResponse(manager, queryPayload, result, logs = "") {
  return {
    success: true,
    framework: manager.framework,
    topic: queryPayload.topic,
    selected_provider: queryPayload.modelIdentifier || queryPayload.provider,
    provider: result.provider,
    model: result.model,
    model_identifier: queryPayload.modelIdentifier || result.modelIdentifier || result.model_identifier || null,
    response: (typeof result.response === "object" && result.response !== null) ? result.response : normalizeResponseText(result.response),
    parameters: {
      temperature: result.temperature,
      max_tokens: result.maxTokens,
      template: queryPayload.template,
    },
    prompt: result.prompt,
    session_id: result.sessionId || queryPayload.sessionId,
    ...(logs ? { debug: logs } : {}),
  };
}

export function serializeDoneEvent(result, sessionId) {
  return {
    type: "done",
    provider: result.provider,
    model: result.model,
    model_identifier: result.modelIdentifier ?? result.model_identifier ?? null,
    response: (typeof result.response === "object" && result.response !== null) ? result.response : null,
    token_usage: result.tokenUsage ?? result.token_usage ?? null,
    session_id: result.sessionId ?? sessionId,
  };
}

export function createWebApi(managerClassOrFactory) {
  const app = createExpressApp();
  let manager;
  const initPromise = (async () => {
    manager = buildManager(managerClassOrFactory);
    await manager._checkProviders();
    return manager;
  })();
  app.use(async (_, __, next) => {
    if (!manager) await initPromise;
    next();
  });

  app.get("/", (_, res) => {
    const available = manager.getAvailableProviders();
    res.json({
      framework: manager.framework,
      available_providers: available,
      available_provider_details: available.map((provider) => buildProviderPayload(provider, manager)),
      available_model_details: modelPayloads(manager),
      total_available: available.length,
      total_models_available: modelPayloads(manager).length,
      initialization_status: manager.initializationMessages,
      status: available.length > 0 ? "healthy" : "no_providers",
    });
  });
  app.get("/providers", (_, res) => {
    const available = manager.getAvailableProviders();
    const models = modelPayloads(manager);
    res.json({
      framework: manager.framework,
      providers: models,
      provider_groups: available.map((provider) => buildProviderPayload(provider, manager)),
      available_providers: available,
      models,
      count: models.length,
      provider_count: available.length,
      model_count: models.length,
      active_provider: models[0]?.name ?? null,
    });
  });
  app.get("/capabilities", (_, res) => {
    res.json({ framework: manager.framework, streaming: true, memory: supportsMemory(manager), memory_retrieval: supportsMemoryRetrieval(manager), coagent: supportsCoagent(manager) });
  });

  app.post("/query", async (req, res) => {
    try {
      const queryPayload = buildQueryPayload(manager, req.body);
      if (!queryPayload.topic) return res.status(400).json({ error: "Topic is required" });
      const { result, logs } = await executeManagerQuery(manager, queryPayload);
      if (result && queryPayload.modelIdentifier) result.modelIdentifier = queryPayload.modelIdentifier;
      if (!resultIsSuccess(result)) return res.status(400).json(queryErrorPayload(result, logs));
      return res.json(serializeQueryResponse(manager, queryPayload, result, logs));
    } catch (error) {
      return res.status(500).json({ error: error.message, framework: manager.framework });
    }
  });

  app.post("/query-stream", async (req, res) => {
    try {
      const queryPayload = buildQueryPayload(manager, req.body);
      if (!queryPayload.topic) return res.status(400).json({ error: "Topic is required" });
      const { result } = await executeManagerQuery(manager, queryPayload);
      res.setHeader("Content-Type", "text/event-stream");
      res.setHeader("Cache-Control", "no-cache");
      res.setHeader("Connection", "keep-alive");
      if (result && queryPayload.modelIdentifier) result.modelIdentifier = queryPayload.modelIdentifier;
      if (!resultIsSuccess(result)) {
        res.write(toSseLine({ type: "error", error: result?.error || "Query failed" }));
        return res.end();
      }
      for await (const event of streamTextSse(normalizeResponseText(result.response), 28, 35)) {
        res.write(event);
      }
      res.write(toSseLine(serializeDoneEvent(result, queryPayload.sessionId)));
      return res.end();
    } catch (error) {
      if (!res.headersSent) return res.status(500).json({ error: error.message, framework: manager.framework });
      res.write(toSseLine({ type: "error", error: error.message }));
      return res.end();
    }
  });

  app.get("/history", async (req, res) => {
    if (!supportsSessionMemory(manager)) return res.status(400).json({ error: "Session memory not supported by this manager" });
    const { provider = "openai", session_id = "default", sessionId = null } = req.query;
    return res.json(await Promise.resolve(manager.getHistory(normalizeProviderInput(manager, provider) || "openai", sessionId ?? session_id ?? "default")));
  });

  app.post("/reset-memory", async (req, res) => {
    if (!supportsSessionMemory(manager)) return res.status(400).json({ error: "Session memory not supported by this manager" });
    const body = req.body || {};
    const provider = body.provider ?? req.query?.provider ?? null;
    const resetSessionId = body.sessionId ?? body.session_id ?? req.query?.sessionId ?? req.query?.session_id ?? null;
    return res.json(await Promise.resolve(manager.resetMemory(normalizeProviderInput(manager, provider), resetSessionId)));
  });

  app.get("/health", (_, res) => {
    const available = manager.getAvailableProviders();
    res.json({ status: available.length > 0 ? "healthy" : "unhealthy", framework: manager.framework, providers_available: available.length });
  });

  return app;
}

export async function runWebServer(managerClassOrFactory, host = "0.0.0.0", port = 8000) {
  const app = createWebApi(managerClassOrFactory);
  let frameworkName = "Unknown";
  try { frameworkName = buildManager(managerClassOrFactory).framework; } catch {}
  app.listen(port, host, () => {
    console.log(`Starting web server for ${frameworkName} framework...`);
    console.log(`Health check: http://${host}:${port}/health`);
    console.log(`Status: http://${host}:${port}/`);
  });
}
